Replacement API for client/v3/naming package to be compatible with new GRPC1.30+ resolver API.

ptabor opened the pull request:
This is not yet implementation, just API definitions.

Please review whether the API is good / can be improved. The goal is to reach agreement about
desired layout of the naming/resolver packages in etcd-3.5+, before investement in the implementation of the API. 

We propose here 3 packages:
  - clientv3/naming/endpoints ->
    That is abstraction layer over etcd that allows to write, read &
watch Endpoints information. It's independent from GRPC API. It hides
the storage details.

  - clientv3/naming/endpoints/internal ->
    That contains the grpc's compatible Update class to preserve the
internal JSON mashalling format.

  - clientv3/naming/resolver ->
    That implements the GRPC resolver API, such that etcd can be
    used for connection.Dial in grpc.

Please see the [grpc_naming.md](TLDR: Please take a look at the proposed modified 3.5 [grpc_naming user guide](https://github.com/etcd-io/etcd/pull/12614/files?short_path=5329e8f#diff-5329e8f4ea030d3052a34c8831bfc948bc5100e935c4cff4175fe75cbd9ee877) document changes & grpcproxy/cluster.go new integration, to see how the new abstractions work.

Note: I'm trying to reach agreement on the vision. I would appreciate (when/if accepted) contributions covering the implementation aspect. 

@skyao @mcluseau @scDisorder @xiang90

scDisorder commented:
for me it looks the way it must to be
and it resolves an issue by the way of grpc documentation
lgtm

gyuho commented on client/v3/naming/endpoints/endpoints.go:
```diff
@@ -0,0 +1,85 @@
+	NewWatcher(ctx context.Context) Watcher
+}
+
+// ManagerAddEndpoint is a syntactic sugar to simplify registration of endpoint in etcd.
```
nice !

ptabor replied:
Thank you.

gyuho commented on client/v3/naming/doc.go:
```diff
@@ -12,45 +12,48 @@
-//		r := &etcdnaming.GRPCResolver{Client: c}
-//		return r.Update(c.Ctx(), service, naming.Update{Op: naming.Add, Addr: addr}, clientv3.WithLease(lid))
+//	func etcdAdd(c *clientv3.Client, lid clientv3.LeaseID, service, addr string) error {
+//    em := endpoints.NewManager(c, service)
```
Can we fix the indentation to be consistent?

ptabor replied:
I wanted to keep the interface minimalistic.
The benefit of `Update` is that is support multiple atomic changes in a single transaction.
`AddEndpoint` is a synthactic sugar on top of Update, so could be implemented in such a way that any implementation of the endpoint.Manager gets it for free. 
It's not that important if we imagine at most 2 implementations: real & fake. 

Let me know - happy to change.

ptabor replied:
Done. (It was tab vs. space issue).

gyuho commented on client/v3/naming/doc.go:
```diff
@@ -12,45 +12,48 @@
-//		r := &etcdnaming.GRPCResolver{Client: c}
-//		return r.Update(c.Ctx(), service, naming.Update{Op: naming.Delete, Addr: "1.2.3.4"})
+//	  em := endpoints.NewManager(c, service)
+//    return endpoints.ManagerDeleteEndpoint(c.Ctx(), em, service+"/"+addr)
```
Make indentation consistent?

ptabor replied:
Done. (It was tab vs. space issue).

gyuho commented on client/v3/naming/doc.go:
```diff
@@ -12,45 +12,48 @@
 //	func etcdAdd(c *clientv3.Client, service, addr string) error {
-//		r := &etcdnaming.GRPCResolver{Client: c}
-//		return r.Update(c.Ctx(), service, naming.Update{Op: naming.Add, Addr: addr})
+//    em := endpoints.NewManager(c, service)
```
Make indentation consistent?

ptabor replied:
Done. (It was tab vs. space issue).

gyuho commented on client/v3/naming/doc.go:
```diff
@@ -12,45 +12,48 @@
-//		etcdnaming "go.etcd.io/etcd/client/v3/naming"
-//
+//  	"go.etcd.io/etcd/client/v3/naming/endpoints"
+//  	"go.etcd.io/etcd/client/v3/naming/resolver"
```
Make indentation consistent?

ptabor replied:
Done

gyuho commented on client/v3/naming/endpoints/endpoints.go:
```diff
@@ -0,0 +1,85 @@
+
+// Manager can be used to add/remove & inspect endpoints stored in etcd.
+type Manager interface {
+	//AddEndpoint(ctx context.Context, key string, endpoint Endpoint, opts ...clientv3.OpOption) error
```
This seems simpler, any specific reason why we don't have this and added `ManagerAddEndpoint`?

ptabor replied:
I wanted to keep the interface minimalistic.
The benefit of Update is that is support multiple atomic changes in a single transaction.
AddEndpoint is a synthactic sugar on top of Update, so could be implemented in such a way that any implementation of the endpoint.Manager gets it for free.
It's not that important if we imagine at most 2 implementations: real & fake.

Let me know - happy to change.

ptabor replied:
Changed as requested.

gyuho commented on client/v3/naming/endpoints/endpoints.go:
```diff
@@ -0,0 +1,85 @@
+// Manager can be used to add/remove & inspect endpoints stored in etcd.
+type Manager interface {
+	//AddEndpoint(ctx context.Context, key string, endpoint Endpoint, opts ...clientv3.OpOption) error
+	//RemoveEndpoint(ctx context.Context, key string) error
```
I think `RemoveEndpoint` as an interface method seems cleaner? Any specific reason we need `ManagerDeleteEndpoint`?

ptabor replied:
ditto

ptabor replied:
Changed as requested.

gyuho commented on client/v3/naming/endpoints/internal/update.go:
```diff
@@ -0,0 +1,37 @@
+package internal
```
Where do we use this package?

ptabor replied:
Inside implementation of the endpoint.Manager. It describes how we serialize and deserialize the endpoints. 

As client/v3 is being is linked into many customer's application, it's important to distinguish public API surface from implementation details. We could keep it lower-case as well, but I wanted in this PR stress its private.

gyuho replied:
I understand that, but `client/v3/naming/endpoints/endpoints_impl.go` importing this looks like placeholders and not implemented yet?

ptabor replied:
Yes. I hope for contributors to implement this within a month. If no one will contribute, I will do the implementation itself. 
The endpoints definition is fixed in stone so its why I'm submitting it.

gyuho commented on server/proxy/grpcproxy/cluster.go:
```diff
@@ -79,38 +74,26 @@ func NewClusterProxy(lg *zap.Logger, c *clientv3.Client, advaddr string, prefix
 
-func (cp *clusterProxy) resolve(prefix string) {
-	rm := rate.NewLimiter(rate.Limit(resolveRetryRate), resolveRetryRate)
-	for rm.Wait(cp.ctx) == nil {
```
So, this deleted part is being moved to `client/v3/naming/endpoints`?

ptabor replied:
I don't think there is anything worth to move in the resolve method.

Please notice that cp.gr.Resolve(prefix) was always succeeding: 

```
func (gr *GRPCResolver) Resolve(target string) (naming.Watcher, error) {	
	ctx, cancel := context.WithCancel(context.Background())	
	w := &gRPCWatcher{c: gr.Client, target: target + "/", ctx: ctx, cancel: cancel}	
	return w, nil	
}
```

So there is no value in the throttling / retry loop. 
That would be needed if we wanted to wait for not empty set of servers..., but it's not the current semantic.

ptabor replied:
Added more explicit error handling.

gyuho commented on client/v3/naming/grpc.go:
```diff
@@ -1,133 +0,0 @@
-// Next gets the next set of updates from the etcd resolver.
-// Calls to Next should be serialized; concurrent calls are not safe since
-// there is no way to reconcile the update ordering.
-func (gw *gRPCWatcher) Next() ([]*naming.Update, error) {
```
So, is this being moved to `client/v3/endpoints/naming`?

ptabor replied:
Yes.  But the API changes slightly and exposes it is a Watcher.Channel(), so it can be easily multiplexed with different 'event' sources.

gyuho reviewed (changes requested):
+1 for the package renaming here. This is mostly moving some of the naming resolution code under `client/v3` -- do I understand right?

Can you add more details regarding my comments in the PR?

ptabor commented:
@gyuha: PTAL

ptabor commented:
@gyuho I changed the PR such that its commitable (I left original naming/grpc.go and introduced the new interfaces side by side, and ["stashed"](https://github.com/ptabor/etcd/commit/142b9a324552e4b859f453ac9ab6dd1a090eb2a6) the grpcproxy migration&cleanup ). 
I propose merging this PR, to unblock other contributors interested in enabling grpc upgrade. 

@scDisorder @dhermes @skyao FYI.

scDisorder commented:
glad to see that! seems like there is not so much to handle ahead

scDisorder commented:
so how it is going? :)

ptabor commented:
@gyuho Are you OK with me merging this ?

gyuho commented on client/v3/naming/endpoints/endpoints_impl.go:
```diff
@@ -0,0 +1,121 @@
+package endpoints
+
+// TODO: The API is not yet implemented.
```
Can we move this as backlogged GitHub issue? Or put these into `internal` package? Once we have this public, it's hard to remove. Seems like this has not been implemented?

ptabor replied:
Added the issue: 
  https://github.com/etcd-io/etcd/issues/12652
I assume this will get implemented as part of 3.5.0 (I will contribute if none else).

gyuho replied:
@ptabor Thanks!

gyuho commented:
> endpoints.NewManager(c, "foo")

This has not been implemented "in" etcd, but I understand this is mainly to unblock gRPC dependency upgrades, so we should be ok to leave it not implemented. Or if we aren't going to have example endpoint manager implementation, can we just remove?
